Move buckets definition to reporter options

nathancyam opened the pull request:
Due to changes on the `telemetry_metrics` repository, which can be found here: https://github.com/beam-telemetry/telemetry_metrics/commit/cc9b30952d36ac8506a2fba30b3905c990de0a7b, the `:buckets` key as part of the `Metrics.Distribution` struct has been dropped. This would cause the call to `distribution` to fail:

```elixir
iex> Metrics.distribution(
  "http.request.duration.seconds",
  buckets: [0.01, 0.025, 0.05, 0.1, 0.2, 0.5, 1],
  event_name: [:http, :request, :complete],
  measurement: :duration,
  unit: {:native, :second}
)
** (KeyError) key :buckets not found in: %{__struct__: Telemetry.Metrics.Distribution, buckeets: [0.01, 0.025, 0.05, 0.1, 0.2, 0.5, 1], description: nil, event_name: [:broadway, :processor, :stop], keep: nil, measurement: #Function<3.23735626/1 in Telemetry.Metrics.maybe_convert_measurement/2>, name: [:broadway, :processor, :stop, :duration, :seconds], reporter_options: [], tag_values: #Function<0.23735626/1 in Telemetry.Metrics.default_metric_options/0>, tags: [], unit: :second}
```

As part of the commit above, it suggests moving the `:buckets` definition to the `reporter_options` keyword, which this PR aims to address.

nathancyam commented on lib/core/registry.ex:
```diff
@@ -204,7 +204,14 @@ defmodule TelemetryMetricsPrometheus.Core.Registry do
+        reporter_options = Keyword.update(
+          metric.reporter_options,
+          :buckets,
+          [],
```
I reckon this could be improved, do you have any ideas @bryannaegele ?

bryannaegele replied:
I don't think we need to provide a default as it's a required option, so we'll instead need to raise an exception during the validation phase. If you remove the empty list here, then this should be fine.

nathancyam replied:
This validation phase, probably would be best fit in `TelemetryMetricsPrometheus.Core.Distribution.register/3`? We have an error case there already, or should we put it in the `Registry.register_metric(%Metrics.Distribution{} = metric, config)` function instead?

nathancyam commented on lib/core/aggregator.ex:
```diff
@@ -31,7 +31,7 @@ defmodule TelemetryMetricsPrometheus.Core.Aggregator do
 
         Enum.sort(samples)
-        |> bucket_measurements(metric.buckets)
+        |> bucket_measurements(metric.reporter_options[:buckets])
```
Personally not a big fan of this.

bryannaegele replied:
This is fine :)

nathancyam commented on lib/core/aggregator.ex:
```diff
@@ -110,7 +110,7 @@ defmodule TelemetryMetricsPrometheus.Core.Aggregator do
   end
 
-  @spec bucket_measurements(measurements :: [number()], buckets :: Metrics.Distribution.buckets()) ::
+  @spec bucket_measurements(measurements :: [number()], buckets :: tuple() | Keyword.t()) ::
```
Is there a better return typespec we could use?

bryannaegele replied:
We are still just passing the buckets here, so it seems like it should still be `[number(), ...]` as seen here https://github.com/beam-telemetry/telemetry_metrics/pull/72/files#diff-bb595d1269de786b707b36f47a7c46a5L34. For convenience, however, it might make sense to just add the `buckets` type to the `Distribution` module.

nathancyam commented:
I've added the bucket validation back into `registry.ex` as well as the unit tests. I've also introduced a new `buckets` type spec as well. I _think_ that's everything. Just a quick question though, how should we version this? Would we call this a breaking change?

bryannaegele commented:
> Would we call this a breaking change?

It is a breaking change. There's at least one other change that will need to be made for the next release. Don't sweat that though. It's expected.

bryannaegele commented on lib/core/registry.ex:
```diff
@@ -81,6 +81,55 @@ defmodule TelemetryMetricsPrometheus.Core.Registry do
+    reporter_options = metric.reporter_options
+
+    unless reporter_options != nil do
+      raise ArgumentError, "expected reporter_options to be on metric"
```
Let's try to provide more info for the user on this error and the buckets one so that the user knows what the error applies to. Maybe something like `reporter_options and the buckets property are required to be defined for Distribution metrics`. Whatever we go with, we should make it clear it applies to the Distribution metric.

nathancyam replied:
Sure thing. Addressed on the latest commit. Error message is: `reporter_options with a valid buckets property are required to be defined for Distribution metrics`. Let me know if it needs a change.
